code/game/objects/obj_defense.dm

```diff
@@ -0,0 +1,236 @@
+
+//the essential proc to call when an obj must receive damage of any kind.
+/obj/proc/take_damage(damage_amount, damage_type = BRUTE, damage_flag = 0, sound_effect = 1, attack_dir)
+	if(sound_effect)
+		play_attack_sound(damage_amount, damage_type, damage_flag)
+	if(!(resistance_flags & INDESTRUCTIBLE) && obj_integrity > 0)
+		damage_amount = run_obj_armor(damage_amount, damage_type, damage_flag, attack_dir)
+		if(damage_amount >= 1)
+			. = damage_amount
+			obj_integrity = max(obj_integrity - damage_amount, 0)
+			if(obj_integrity <= 0)
+				obj_destruction(damage_flag)
+			else if(integrity_failure)
+				if(obj_integrity <= integrity_failure)
+					obj_break(damage_flag)
+
+//returns the damage value of the attack after processing the obj's various armor protections
+/obj/proc/run_obj_armor(damage_amount, damage_type, damage_flag = 0, attack_dir)
+	switch(damage_type)
+		if(BRUTE)
+		if(BURN)
+		else
+			return 0
+	var/armor_protection = 0
+	if(damage_flag)
+		armor_protection = armor[damage_flag]
+	return round(damage_amount * (100 - armor_protection)*0.01, 0.1)
+
+//the sound played when the obj is damaged.
+/obj/proc/play_attack_sound(damage_amount, damage_type = BRUTE, damage_flag = 0)
+	switch(damage_type)
+		if(BRUTE)
+			if(damage_amount)
+				playsound(src, 'sound/weapons/smash.ogg', 50, 1)
+			else
+				playsound(src, 'sound/weapons/tap.ogg', 50, 1)
+		if(BURN)
+			playsound(src.loc, 'sound/items/Welder.ogg', 100, 1)
+
+/obj/hitby(atom/movable/AM)
+	..()
+	var/tforce = 0
+	if(ismob(AM))
+		tforce = 10
+	else if(isobj(AM))
+		var/obj/O = AM
+		tforce = O.throwforce
+	take_damage(tforce, BRUTE, "melee", 1, get_dir(src, AM))
+
+/obj/ex_act(severity, target)
+	..() //contents explosion
+	if(target == src)
+		qdel(src)
+		return
+	switch(severity)
+		if(1)
+			qdel(src)
+		if(2)
+			take_damage(rand(100, 250), BRUTE, "bomb", 0)
+		if(3)
+			take_damage(rand(10, 90), BRUTE, "bomb", 0)
+
+/obj/bullet_act(obj/item/projectile/P)
+	. = ..()
+	visible_message("<span class='danger'>[src] is hit by \a [P]!</span>")
+	playsound(src, P.hitsound, 50, 1)
+	take_damage(P.damage, P.damage_type, P.flag, 0, turn(P.dir, 180))
+
+
+/obj/attack_hulk(mob/living/carbon/human/user, does_attack_animation = 0)
+	if(user.a_intent == "harm")
+		..(user, 1)
+		visible_message("<span class='danger'>[user] smashes [src]!</span>")
+		if(density)
+			playsound(src, 'sound/effects/meteorimpact.ogg', 100, 1)
+			user.say(pick(";RAAAAAAAARGH!", ";HNNNNNNNNNGGGGGGH!", ";GWAAAAAAAARRRHHH!", "NNNNNNNNGGGGGGGGHH!", ";AAAAAAARRRGH!" ))
+		else
+			playsound(src, 'sound/effects/bang.ogg', 50, 1)
+		take_damage(150, BRUTE, "melee", 0, get_dir(src, user))
+		return 1
+	return 0
+
+/obj/blob_act(obj/structure/blob/B)
+	take_damage(400, BRUTE, "melee", 0, get_dir(src, B))
```

Comment:
Huge damage amount (exception for mecha for balance purpose) because the blob pretty much destroy everything in its path.